llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -2428,6 +2428,15 @@ InnerLoopVectorizer::getOrCreateVectorTripCount(BasicBlock *InsertBlock) {
   Value *TC = getTripCount();
   IRBuilder<> Builder(InsertBlock->getTerminator());
 
+  // Use original trip count as the vector trip count if use predicated EVL
+  // instructions for tail-folding.
+  if (VF.isVector() && Cost->foldTailWithEVL()) {
+    assert(!Cost->requiresScalarEpilogue(true) &&
+           "Use predicated EVL instructions for tail-folding does not allow "
+           "scalar epilogue");
+    return VectorTripCount = TC;
+  }
+
```

Comment:
This doesn't look like it is the right place to update. The vector trip count is guaranteed to be a multiple of VF.

The caller would need updating to not use it/request when folding the tail with EVL